staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go

```diff
@@ -423,20 +398,293 @@ func TestPriorityAndFairnessWithPanicRecoverAndTimeoutFilter(t *testing.T) {
+		})
+		handler := newHandlerChain(t, requestHandler, controller, userName, requestTimeout)
+
+		server, requestGetter := newHTTP2ServerWithClient(handler)
+		defer server.Close()
+
+		response, err := requestGetter(rquestTimesOutPath)
+		if err != nil {
+			t.Fatalf("expected request: %s to succeed, but got error: %#v", rquestTimesOutPath, err)
+		}
+		if response.StatusCode != http.StatusGatewayTimeout {
+			t.Errorf("expected HTTP status code: %d for request: %s, but got: %#v", http.StatusGatewayTimeout, rquestTimesOutPath, response)
+		}
+
+		close(stopCh)
+		t.Log("waiting for the controller to shutdown")
+
+		controllerErr := <-controllerCompletedCh
+		if controllerErr != nil {
+			t.Errorf("expected a nil error from controller, but got: %#v", controllerErr)
+		}
+	})
+
+	t.Run("priority level concurrency is set to 1, inner handler writes to the response before request times out", func(t *testing.T) {
+		const (
+			requestTimeout                                        = 1 * time.Second
+			userName                                              = "alice"
+			fsName                                                = "test-fs"
+			plName                                                = "test-pl"
+			serverConcurrency, plConcurrencyShares, plConcurrency = 1, 1, 1
+		)
+
+		apfConfiguration := newConfiguration(fsName, plName, userName, plConcurrencyShares, 0)
+		stopCh := make(chan struct{})
+		controller, controllerCompletedCh := startAPFController(t, stopCh, apfConfiguration, serverConcurrency, requestTimeout/4, plName, plConcurrency)
+
+		rquestTimesOutPath := "/request/time-out"
+		ctx, cancel := context.WithTimeout(context.Background(), requestTimeout+time.Second)
+		defer cancel()
+		requestHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			expectMatchingAPFHeaders(t, w, fsName, plName)
+
+			if r.URL.Path == rquestTimesOutPath {
+				// inner handler writes header and then let the request time out.
+				w.WriteHeader(http.StatusBadRequest)
+				<-ctx.Done()
+
+				// we expect the timeout handler to have timed out this request by now and any attempt
+				// to write to the response should return a http.ErrHandlerTimeout error.
+				if _, err := w.Write([]byte("foo")); err != http.ErrHandlerTimeout {
+					t.Fatalf("expected error: %#v, but got: %#v", http.ErrHandlerTimeout, err)
+				}
+			}
+		})
+		handler := newHandlerChain(t, requestHandler, controller, userName, requestTimeout)
+
+		server, requestGetter := newHTTP2ServerWithClient(handler)
+		defer server.Close()
+
+		_, err := requestGetter(rquestTimesOutPath)
+		expectResetStreamError(t, err)
+
+		close(stopCh)
+		t.Log("waiting for the controller to shutdown")
+
+		controllerErr := <-controllerCompletedCh
+		if controllerErr != nil {
+			t.Errorf("expected a nil error from controller, but got: %#v", controllerErr)
+		}
+	})
+
+	t.Run("priority level concurrency is set to 1, queue length is 1, enqueued request should be rejected", func(t *testing.T) {
```

Comment:
@MikeSpreitzer queue length=1, I have two concurrent requests. The first one is being executed and and is timed out by design. This ensures that the second request stays en-queued. Since, there is not a third request to come and do
```
	// The next step is the logic to reject requests that have been waiting too long
	qs.removeTimedOutRequestsFromQueueLocked(queue, fsName)
```

The second request actually times out with a 504 to client.

```
	doneCh := ctx.Done()
	if doneCh != nil {
		qs.preCreateOrUnblockGoroutine()
		go func() {
			defer runtime.HandleCrash()
			qs.goroutineDoneOrBlocked()
			_ = <-doneCh
			// Whatever goroutine unblocked the preceding receive MUST
			// have already either (a) incremented qs.counter or (b)
			// known that said counter is not actually counting or (c)
			// known that the count does not need to be accurate.
			// BTW, the count only needs to be accurate in a test that
			// uses FakeEventClock::Run().
			klog.V(6).Infof("QS(%s): Context of request %q %#+v %#+v is Done", qs.qCfg.Name, fsName, descr1, descr2)
			qs.cancelWait(req)
			qs.goroutineDoneOrBlocked()
		}()
	}
	return req, false
```
The context used by the go routine  above is the one created by the timeout handler which has a fixed `60s` timeout. Please note it is not the timeout specified by the user. I think we can use a context here that reflects the actual deadline of the request, or one created from it with a shorter (1/4) deadline.

May be when https://github.com/kubernetes/kubernetes/pull/96901 merges it will be easier. Let me know your thoughts.

Reply:
In case it is not clear, this test makes an assumption that is plausible but untrue and not explicitly discussed in the comments on the relevant functions.  The assumption is that if and when a request that is executing (not waiting in a queue) times out, the APF layer considers the request to have finished executing.  I initially coded it that way, but during review the decision was made to have the APF layer consider the request to be executing until the inner handler actually returns.  #96901 may make that happen promptly in more cases than happens now, but will probably not cover all cases.

Thus, in the scenario tested by this test, the first request executes for longer than the request timeout, and this blocks the second request from starting for that long, so the second request is also reaped by the timeout filter.  We made an explicit decision to check for timeouts on waiting requests only in the course of processing another request; we did this for the sake of keeping down the code and runtime complexity, figuring it would be a good enough approximation of the ideal.